### PR TITLE
[core] refactor `step` method

### PR DESCRIPTION
**Pull request by younesbelkada**
This PR adds a new safety checker inside `step` method to make sure the rewards are set on the correct device. Regarding the queries and the responses users should retrieve the dataloader from the trainer and use that instead as the device assignment is performed directly at the dataloader level. Since the reward is not part of the dataloader the device assignment needs to be performed manually.

This PR also adds inside the safety checker a new check. Before this PR if a user pass a reward tensor with a dimension different from 0 (e.g. `torch.tensor([1.0])`) it would break the training loop. Therefore now we force the reward tensor to be with the desired shape. We now also throw a value error if the dimension of the reward is > 1.

cc @lewtun @lvwerra @edbeeching

### CODE REVIEW

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -252,6 +255,17 @@ def _step_safety_checker(
                     f"Batch size ({batch_size}) does not match number of examples - but got {len(tensor_list)} for: {name}"
                 )
 
+            # set scores on the correct device
+            if name == "scores":
+                scores = [score.to(self.accelerator.device) for score in scores]
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:260]: I think we can do the same for queries and values as well. although they should be you don't know what the user might do before passing them to `step` and then the behaviour is consistent for all inputs. 

Do you know what happens when the tensor is already on device? Will it copy it again or do nothing?

**Review by lvwerra — COMMENTED** [review]
Generally looks good to me, just one comment! 🚀

File: trl/trainer/ppo_trainer.py (hunk 1 above)
Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py:260]: Makes sense, in PT it should do nothing if they are on the same device so this should be cost-free

Comment (younesbelkada) [reply trl/trainer/ppo_trainer.py]: Should be now handled in [b32bfbf](https://github.com/lvwerra/trl/pull/76/commits/b32bfbffaa16abde73e12ba838c311dcfbbd175b)

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -252,6 +255,25 @@ def _step_safety_checker(
...
+            else:
+                scores = tensor_list
+                # squeeze scores if needed
+                for i, score in enumerate(scores):
+                    if score.dim() > 1:
+                        raise ValueError(f"Scores must be 1-dimensional - got {score.dim()} for {score}")
+                    elif score.dim() == 1:
+                        scores[i] = score.squeeze()
+
+        return queries, responses, scores
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:275]: What do you think about that simplification?

```suggestion
        # add queries, scores and responses on the correct device
        queries = [tensor.to(self.accelerator.device) for tensor in queries]
        responses = [tensor.to(self.accelerator.device) for tensor in responses]
        scores = [tensor.to(self.accelerator.device) for tensor in scores]
    
        # squeeze scores if needed
        for i, score in enumerate(scores):
            if score.dim() > 1:
                raise ValueError(f"Scores must be 1-dimensional - got {score.dim()} for {score}")
            elif score.dim() == 1:
                    scores[i] = score.squeeze()

        return queries, responses, scores
```

File: tests/test_ppo_trainer.py [hunk 3]
```diff
@@ -230,3 +230,42 @@ def test_ppo_step_with_ref_and_custom_layers_warning(self):
                 dataset=dummy_dataset,
                 num_shared_layers=num_shared_layers,
             )
+
+    def test_ppo_step_rewards_shape(self):
```
Comment (lvwerra) [inline tests/test_ppo_trainer.py:234]: can we also add a test that checks the query/response shape?

Comment (younesbelkada) [reply tests/test_ppo_trainer.py:234]: Addressed in [e3d4e26](https://github.com/lvwerra/trl/pull/76/commits/e3d4e26e118baf52bebe6a42b0285447ab17505b)

**Review by lvwerra — APPROVED** [review]
LGTM!
